Allow safer hooking of OPENSSL_cleanup in providers

levitte opened the pull request:
Calling OPENSSL_cleanup() through atexit() from a libcrypto that's loaded via a provider module is a bad idea, since the handler will most likely be called after the provider has been unloaded.

This adds a mechanism that allows a provider to hook the ~addition of atexit functions~ libcrypto cleanup via an upcall that's offered by the calling libcrypto, and have an internal mechanism that ensures that the cleanup isn't called too early.  The legacy provider module is enhanced with this.

Because the FIPS provider only contains a subset of libcrypto, and nothing that gives cause to call atexit(), we don't change it.

levitte committed:
Revert "Statically link the legacy provider to endecode_test"

Reverts commit 7ee992a5d931ab5ad9df00d2d8e47e1b7a72d7ac

This is not necessary any more, as using the ERR upcalls renders this
hack unnecessary.

levitte committed:
PROV: implement a cleanup hook for providers

This provides an upcall, OSSL_FUNC_core_hook_cleanup, which stores the
passed function as an extra cleanup function to be performed on provider
teardown, as part of ossl_provider_free(), directly after the call of
the teardown function.

levitte committed:
INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

This involves passing the OSSL_CORE_HANDLE and OSSL_DISPATCH that
OSSL_provider_init() receives via an allocated OPENSSL_INIT_SETTINGS.

To ensure that all cleanup hooks are accounted for, we register the
OSSL_CORE_HANDLE handles, and when the cleanup function is invoked,
these handles are unregistered, and OPENSSL_cleanup() is called when
the set of registered handles is empty.

levitte committed:
Use OPENSSL_init_crypto(OPENSSL_INIT_PROVIDER, settings) in legacy provider

levitte commented:
Currently a draft, as this is unfinished.

Among other things, we must ensure that multiple provider modules that share the same libcrypto in process space don't cause a cleanup too early.  This is probably best handled in libcrypto itself...  how remains to be seen.

levitte commented:
Before anyone asks, I don't consider this to be 3.0 material

rsbeckerca commented:
@levitte which thread of this would you like tested on NonStop (both is ok, just takes longer)?

rsbeckerca commented:
> > @levitte which thread of this would you like tested on NonStop (both is ok, just takes longer)?
> 
> Not sure what you mean with "thread", but it would be very nice if you were willing to give this PR a try!

Ok. Will finish the test cycle on 9c31ab8ade (legacy-provider-use-error-upcalls) mostly because it is 90% done, then move on to this PR.

mattcaswell commented:
> Among other things, we must ensure that multiple provider modules that share the same libcrypto in process space don't cause a cleanup too early. This is probably best handled in libcrypto itself... how remains to be seen.

Another related problem is the issue of the *same* provider module loaded from different libctxs. E.g. consider

````
    OSSL_LIB_CTX *ctx1 = OSSL_LIB_CTX_new();
    OSSL_LIB_CTX *ctx2 = OSSL_LIB_CTX_new();

    OSSL_PROVIDER *prov1 = OSSL_PROVIDER_load(ctx1, "myprovider");
    OSSL_PROVIDER *prov2 = OSSL_PROVIDER_load(ctx2, "myprovider");

    /* Use prov1 */

    OSSL_PROVIDER_unload(prov1);
    OSSL_LIB_CTX_free(ctx1);

    /* OPENSSL_cleanup() has now been called for myprovider.so */

    /* Use prov2....bang */
````

levitte commented:
> Among other things, we must ensure that multiple provider modules that share the same libcrypto in process space don't cause a cleanup too early. This is probably best handled in libcrypto itself... how remains to be seen.

That's probably exactly what causes the `test_evp_kdf` test to fail in all shared configurations

rsbeckerca commented:
> > > @levitte which thread of this would you like tested on NonStop (both is ok, just takes longer)?
> > 
> > 
> > Not sure what you mean with "thread", but it would be very nice if you were willing to give this PR a try!
> 
> Ok. Will finish the test cycle on [9c31ab8](https://github.com/openssl/openssl/commit/9c31ab8aded7766bbffd18f73af15c361077bdb5) (legacy-provider-use-error-upcalls) mostly because it is 90% done, then move on to this PR.

Starting test of this PR on NonStop x86

rsbeckerca commented:
This may be a separate issue. Here is the trace of the above failure:
```
#0  0x7fc52ef9 in CRYPTO_THREAD_read_lock (lock=0x0)
    at /home/ituglib/randall/openssl-3.0/crypto/threads_none.c:36
#1  0x7fc2f57b in ossl_lib_ctx_get_data (ctx=0x7fff7100, index=0,
    meth=0x7ff7c5c0) at /home/ituglib/randall/openssl-3.0/crypto/context.c:360
#2  0x7fbe09ac in get_evp_method_store (libctx=0x0)
    at /home/ituglib/randall/openssl-3.0/crypto/evp/evp_fetch.c:82
#3  0x7fbe1133 in inner_evp_generic_fetch (methdata=0x6fffe9f0, prov=0x0,
    operation_id=4, name_id=0, name=0x80010a0 "KBKDF", properties=0x0,
    new_method=0x7fbf0220 <evp_kdf_from_algorithm>,
    up_ref_method=0x7fbf0000 <evp_kdf_up_ref>,
    free_method=0x7fbf0080 <evp_kdf_free>)
    at /home/ituglib/randall/openssl-3.0/crypto/evp/evp_fetch.c:248
#4  0x7fbe1925 in evp_generic_fetch (libctx=0x0, operation_id=4,
    name=0x80010a0 "KBKDF", properties=0x0,
    new_method=0x7fbf0220 <evp_kdf_from_algorithm>,
    up_ref_method=0x7fbf0000 <evp_kdf_up_ref>,
    free_method=0x7fbf0080 <evp_kdf_free>)
    at /home/ituglib/randall/openssl-3.0/crypto/evp/evp_fetch.c:372
#5  0x7fbf0806 in EVP_KDF_fetch (libctx=0x0, algorithm=0x80010a0 "KBKDF",
    properties=0x0)
    at /home/ituglib/randall/openssl-3.0/crypto/evp/kdf_meth.c:163
#6  0x70005856 in get_kdfbyname_libctx (libctx=0x0, name=0x80010a0 "KBKDF")
    at /home/ituglib/randall/openssl-3.0/test/evp_kdf_test.c:25
#7  0x700058c5 in get_kdfbyname (name=0x80010a0 "KBKDF")
    at /home/ituglib/randall/openssl-3.0/test/evp_kdf_test.c:34
#8  0x7000b5f9 in test_kdf_kbkdf_6803_128 ()
    at /home/ituglib/randall/openssl-3.0/test/evp_kdf_test.c:989
#9  0x700111ed in run_tests (
    test_prog_name=0x800d010 "../../test/evp_kdf_test")
    at /home/ituglib/randall/openssl-3.0/test/testutil/driver.c:334
#10 0x70011d23 in main (argc=1, argv=0x800d000)
    at /home/ituglib/randall/openssl-3.0/test/testutil/main.c:30
#11 0x700057e0 in _MAIN () at \BLPROD.$RTAL.T8432ACM.CMAIN64C:52
```
Although a null `libctx`?

levitte commented:
What you're experiencing (and I do too, exactly the same way) is exactly what I commented higher up, quoted here:

> Among other things, we must ensure that multiple provider modules that share the same libcrypto in process space don't cause a cleanup too early. This is probably best handled in libcrypto itself... how remains to be seen.

OPENSSL_cleanup was called too early, because libcrypto is referred to more than once, meaning that the remaining "instance" doesn't have the resources it expects (locks etc) any longer. Kaboom.

I have some loose ideas how to fix that, I just need to let them simmer overnight.

levitte commented:
A null libctx is normal, that's the default one

rsbeckerca commented:
> What you're experiencing (and I do too, exactly the same way) is exactly what I commented higher up, quoted here:
> 
> > Among other things, we must ensure that multiple provider modules that share the same libcrypto in process space don't cause a cleanup too early. This is probably best handled in libcrypto itself... how remains to be seen.
> 
> OPENSSL_cleanup was called too early, because libcrypto is referred to more than once, meaning that the remaining "instance" doesn't have the resources it expects (locks etc) any longer. Kaboom.
> 
> I have some loose ideas how to fix that, I just need to let them simmer overnight.

Simmering is good. I will monitor and test as commits happen. Thanks. Your efforts are helping to sustain the relevance of this product on my platform.

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

levitte commented:
@rsbeckerca, I believe (or hope) that I've finally made something somewhat sensible.  mind giving this PR a try now?

levitte commented:
@mattcaswell, a note: I moved down the removal of `init_lock` in `OPENSSL_cleanup()`...  would you mind double checking for me if that could cause any disruptions or unintended freezes?  I don't think it should, but...

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

levitte commented:
Looking at this a little more, I realised that it's pretty simple to remove our use of atexit() in library...  like "just do it" kind of easy!  It does mean that applications will have to start calling OPENSSL_cleanup() explicitly, so I still don't think that's feasible to do in 3.x.

However, in the interest of actually trying that out, this is the patch I came up with:

``` diff
diff --git a/apps/openssl.c b/apps/openssl.c
index 3497d51f4d..63e3ad4266 100644
--- a/apps/openssl.c
+++ b/apps/openssl.c
@@ -308,6 +308,7 @@ int main(int argc, char *argv[])
     BIO_free_all(bio_out);
     apps_shutdown();
     BIO_free(bio_err);
+    OPENSSL_cleanup();
     EXIT(ret);
 }
 
diff --git a/crypto/init.c b/crypto/init.c
index 3352f62d58..f9315f0f13 100644
--- a/crypto/init.c
+++ b/crypto/init.c
@@ -197,43 +197,12 @@ static void ossl_init_do_cleanup(const OSSL_CORE_HANDLE *handle)
 }
 
 static CRYPTO_ONCE register_atexit = CRYPTO_ONCE_STATIC_INIT;
-#if !defined(OPENSSL_SYS_UEFI) && defined(_WIN32)
-static int win32atexit(void)
-{
-# ifdef OPENSSL_INIT_DEBUG
-    fprintf(stderr,
-            "OPENSSL_INIT: win32atexit(): running OPENSSL_cleanup()\n");
-# endif
-    OPENSSL_cleanup();
-    return 0;
-}
-#else
-static void stdatexit(void)
-{
-# ifdef OPENSSL_INIT_DEBUG
-    fprintf(stderr,
-            "OPENSSL_INIT: stdatexit(): running OPENSSL_cleanup()\n");
-# endif
-    OPENSSL_cleanup();
-}
-#endif
-
-DEFINE_RUN_ONCE_STATIC(ossl_init_register_atexit)
+DEFINE_RUN_ONCE_STATIC(ossl_init_no_register_atexit)
 {
 #ifdef OPENSSL_INIT_DEBUG
-    fprintf(stderr, "OPENSSL_INIT: ossl_init_register_atexit()\n");
-#endif
-#ifndef OPENSSL_SYS_UEFI
-# if defined(_WIN32) && !defined(__BORLANDC__)
-    /* We use _onexit() in preference because it gets called on DLL unload */
-    if (_onexit(win32atexit) == NULL)
-        return 0;
-# else
-    if (atexit(stdatexit) != 0)
-        return 0;
-# endif
+    fprintf(stderr, "OPENSSL_INIT: ossl_init_no_register_atexit ok!\n");
 #endif
-
+    /* Do nothing in this case */
     return 1;
 }
 
@@ -242,7 +211,7 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_register_atexit)
  * the only difference being the debug message.
  */
 DEFINE_RUN_ONCE_STATIC_ALT(ossl_init_register_provider_atexit,
-                           ossl_init_register_atexit)
+                           ossl_init_no_register_atexit)
 {
 #ifdef OPENSSL_INIT_DEBUG
     fprintf(stderr, "OPENSSL_INIT: ossl_init_register_provider_atexit()\n");
@@ -251,16 +220,6 @@ DEFINE_RUN_ONCE_STATIC_ALT(ossl_init_register_provider_atexit,
     return 1;
 }
 
-DEFINE_RUN_ONCE_STATIC_ALT(ossl_init_no_register_atexit,
-                           ossl_init_register_atexit)
-{
-#ifdef OPENSSL_INIT_DEBUG
-    fprintf(stderr, "OPENSSL_INIT: ossl_init_no_register_atexit ok!\n");
-#endif
-    /* Do nothing in this case */
-    return 1;
-}
-
 int OPENSSL_INIT_set_upcalls(OPENSSL_INIT_SETTINGS *settings,
                              const OSSL_CORE_HANDLE *handle,
                              const OSSL_DISPATCH *upcalls)
@@ -706,11 +665,7 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
      * and leave it to the OPENSSL_cleanup() call which has been registered
      * through atexit(), or that the application does explicitly.
      */
-    if ((opts & OPENSSL_INIT_NO_ATEXIT) != 0) {
-        if (!RUN_ONCE_ALT(&register_atexit, ossl_init_no_register_atexit,
-                          ossl_init_register_atexit))
-            return 0;
-    } else if ((opts & OPENSSL_INIT_PROVIDER) != 0) {
+    if ((opts & OPENSSL_INIT_PROVIDER) != 0) {
         int ret;
         int (*c_hook_cleanup)(const OSSL_CORE_HANDLE *handle,
                               void (*cleanup_function)(const OSSL_CORE_HANDLE
@@ -725,7 +680,7 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
             return 0;
         ret = RUN_ONCE_ALT(&register_atexit,
                            ossl_init_register_provider_atexit,
-                           ossl_init_register_atexit);
+                           ossl_init_no_register_atexit);
 
         if (ret <= 0)
             return 0;
@@ -743,7 +698,11 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
             return 0;
         ossl_init_register_cleanup(settings->handle);
     } else {
-        if (!RUN_ONCE(&register_atexit, ossl_init_register_atexit))
+        /*
+         * NO_ATEXIT is default, and assumes that the application calls
+         * the cleanup.
+         */
+        if (!RUN_ONCE(&register_atexit, ossl_init_no_register_atexit))
             return 0;
         ossl_init_register_cleanup(NULL);
     }
```

rsbeckerca commented:
Will test when I'm awake :)

levitte commented:
> Good news. All tests passed on NonStop x86 at commit [19ab700](https://github.com/openssl/openssl/commit/19ab700fb0fa9fbfbaa7e4aff4950073b6de739a).

Yay!

levitte commented:
With that, this is no longer a draft

levitte commented on crypto/init.c:
```diff
@@ -78,13 +78,144 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_base)
+                ": running OPENSSL_cleanup()\n",
+                (void *)handle);
+#endif
+        OPENSSL_cleanup();
```
I have ignored engines in this PR.  That should be easy enough by having an upcall similar to `OSSL_FUNC_core_hook_cleanup` in `dynamic_fns`, and do the exact same `OPENSSL_init_crypto(OPENSSL_INIT_PROVIDER, ...)` dance in `IMPLEMENT_DYNAMIC_BIND_FN` as legacyprov.c does in this PR.

Follow-up PR welcome ;-)

levitte replied:
I've thought a bit more on this, and realised it might not be as simple as I made it out to be...  Still worth a thought, though...

yanovich replied:
The whole idea of automatically unloading `libcrypto` doesn't seem to be worth the effort. Most popular use cases for `libcrypto` are a command-line `opensll`, `curl`/`wget` invocations, and browser/server SSL/TLS handling. If used in a CLI app, `libcrypto` will be unloaded when the app finishes its task and terminates. A long-running app like a browser or a server usually loads its modules at start-up and keeps them loaded until terminated.

All other uses are edge cases. It is perfectly reasonable just to provide a method to manually unload the library. All other optimizations could be done by those who need them either in client code or as PRs.

levitte commented:
Apart from engines not being covered here, final review? Ping?

paulidale commented on crypto/init.c:
```diff
@@ -78,13 +78,144 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_base)
+     * Trust that the memory allocator doesn't actually reallocate
+     * when the size is unchanged.
+     */
+    handles
```
Why can't you use a stack for this?
It seems like you should be using one.

levitte replied:
Lots of extra churn.  Also, slight paranoia...  the init code is quite sensitive, I didn't want to add too much dependencies on the rest of libcrypto, on the off chance that we make changes in, for example, the stack code that wouldn't sit well here.

paulidale replied:
I still think it would be better to use the well tested stack code rather than hacking something together here -- the comment about trusting the memory allocator indicates the level of hack.  A realloc implementation that always allocates a new block and uses memcpy to duplicate the final size is complaint with the standard.

If you want to avoid the stack code, at least use a singly linked list and allocate each node separately.  What's here is just bad.

levitte replied:
Actually, that comment is unnecessary, it's not really a crucial thing here.

A singly linked list seems like a lot of churn for a very simple array of addresses.

I'll start with removing that comment.

t8m replied:
The realloc can fail and this will leak the original handles allocation.

paulidale replied:
Just use a stack, don't re-implement things unnecessarily.

levitte replied:
Either way, a decision needs to be made about what to do when allocation fails.  Currently, it happens "naturally", by crashing of `handles` ends up being NULL.  Whatever we do, if allocation fails, I can't see how to continue, as this array is essentially corrupt.  Changing this to a stack will make no difference in that case, except having to add some sort of explicit crash code.

levitte replied:
The stack code calls ERR_raise in some places...  that's the actual call that got us into this journey.  Calling something that leads there here has all my internal warning claxons go off at the same time.

So no, will not use the stack.

paulidale replied:
Have a static array with, say, 100 elements.  No malloc/realloc no possible failure except overflow.

levitte replied:
Okie, done

levitte commented:
Rebased on top of current master

paulidale commented on doc/man7/provider.pod:
```diff
@@ -72,6 +72,39 @@ the provider.
+made into dynamically loadable modules and for providers that are built into the
+application.
+
+=begin comment
```
This shouldn't be in a comment.  A comment is a guarantee that nobody will ever read it.  The text should be included in the man page or removed entirely.  In the latter case, they could be included in a C source file as a comment.

I think they contain useful information and could be included here.

levitte replied:
That information doesn't really concern external people that want to write a provider.

So, where would you place those comments in code?  I've tried to figure out where I would be most likely to go looking if I was wondering, and this is what I came up with, se re "nobody will ever read it", I'm apparently nobody...

levitte replied:
I may have found another way, so fix coming up

levitte replied:
Done

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

levitte committed:
Add a comment in our other providers regarding OPENSSL_init_crypto()

In those providers, OPENSSL_init_crypto(OPENSSL_INIT_PROVIDER, ...)
isn't necessary for diverse reasons...  in the FIPS module, it's not
even possible to use that mechanism.

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

rsbeckerca commented:
FYI: all tests pass when 63bf7df4b6 is merged onto 14951ef01f.

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

paulidale commented on crypto/init.c:
```diff
@@ -78,13 +78,137 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_base)
+        goto end;
+
+    ret = -1;
+    if (handles_num < OSSL_NELEM(handles)) {
```
What about making this an ossl_assert?
It will make finding an overflow a lot easier.

levitte replied:
Sure

levitte replied:
Done

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

kaduk commented on providers/fips/fipsprov.c:
```diff
@@ -545,6 +545,14 @@ int OSSL_provider_init_int(const OSSL_CORE_HANDLE *handle,
 
+    /*
+     * The OPENSSL_init_crypto(OPENSSL_INIT_PROVIDER, ...) call that's
+     * indicated in provider(7) isn't necessary here, since The FIPS module
```
```suggestion
     * indicated in provider(7) isn't necessary here, since the FIPS module
```

levitte replied:
Fixed

levitte committed:
fixup! Add a comment in our other providers regarding OPENSSL_init_crypto()

mattcaswell commented on crypto/init.c:
```diff
@@ -521,15 +679,66 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
+    } else {
+        if (!RUN_ONCE(&register_atexit, ossl_init_register_atexit))
+            return 0;
+        ossl_init_register_cleanup(NULL);
```
Since this is outside of the RUN_ONCE, is there a possible race condition here?

levitte replied:
Not sure why there would be.  The core provider code runs OPENSSL_init_crypto() before loading anything, and this line gets executed before OPENSSL_INIT_LOAD_CONFIG is processed, so you tell me.

mattcaswell commented on crypto/init.c:
```diff
@@ -521,15 +679,66 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
+     * torn down.
+     * Care is taken to handle the case where libcrypto is shared by multiple
+     * providers, by registering the provider handle.  When the cleanup
+     * function is called, that same handle is unregistered, when when the
```
when when

levitte replied:
Fixed

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -683,6 +686,8 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
             if (prov->flag_initialized) {
                 ossl_provider_teardown(prov);
+                if (prov->cleanup_function != NULL)
+                    prov->cleanup_function((OSSL_CORE_HANDLE *)prov);
```
I'm struggling to see the difference between the provider's teardown function, and the provider's cleanup_function?

mattcaswell reviewed (changes requested):
I am not entirely convinced that this PR is quite correct.

IIUC, it assumes that the when the last active provider has been unloaded then it is safe to call OPENSSL_cleanup. But it is only safe to call OPENSSL_cleanup if no more providers are loaded before the DSO is removed from memory. If a provider is subsequently loaded after OPENSSL_cleanup has been called, but before the DSO is removed from memory then it looks to me like this is going to go wrong.

levitte committed:
fixup! INIT: Add the possibility to hook the cleanup upcall via OPENSSL_init_crypto()

mattcaswell commented:
> I think I understand what you mean. The last provider sharing a subordinate libcrypto has called OPENSSL_cleanup(), and then another thread of the calling application loads another provider sharing the same libcrypto.

Yes. Although it doesn't have to be a different thread. Something like this:

````
/* Simulate the OS keeping the DSO around for longer than we might expect... */
legacydso = DSO_load(NULL, "legacy.so", NULL, 0);
ctx = OSSL_LIB_CTX_new();
legacy = OSSL_PROVIDER_load("legacy");
/* Do something with legacy that causes OPENSSL_init_crypto to run */
OSSL_PROVIDER_unload(legacy);
OSSL_LIB_CTX_free(ctx);
/* If legacy.so is using a different libcrypto to the application then legacy's OPENSSL_cleanup() now gets run */
/* Now reload legacy */
ctx = OSSL_LIB_CTX_new();
legacy = OSSL_PROVIDER_load("legacy");
/* Do something with legacy that causes OPENSSL_init_crypto to run */
/* This is going to fail because OPENSSL_cleanup() was run */
````

yanovich commented:
> I think I understand what you mean. The last provider sharing a subordinate libcrypto has called OPENSSL_cleanup(), and then another thread of the calling application loads another provider sharing the same libcrypto.

This case can probably be ignored, if the program is in the `atexit` phase of execution.

levitte commented:
> > I think I understand what you mean. The last provider sharing a subordinate libcrypto has called OPENSSL_cleanup(), and then another thread of the calling application loads another provider sharing the same libcrypto.
> 
> Yes. Although it doesn't have to be a different thread. Something like this:
> 
> ```
> /* Simulate the OS keeping the DSO around for longer than we might expect... */
> legacydso = DSO_load(NULL, "legacy.so", NULL, 0);
> ctx = OSSL_LIB_CTX_new();
> legacy = OSSL_PROVIDER_load("legacy");
> /* Do something with legacy that causes OPENSSL_init_crypto to run */
> OSSL_PROVIDER_unload(legacy);
> OSSL_LIB_CTX_free(ctx);
> /* If legacy.so is using a different libcrypto to the application then legacy's OPENSSL_cleanup() now gets run */
> /* Now reload legacy */
> ctx = OSSL_LIB_CTX_new();
> legacy = OSSL_PROVIDER_load("legacy");
> /* Do something with legacy that causes OPENSSL_init_crypto to run */
> /* This is going to fail because OPENSSL_cleanup() was run */
> ```

Have you *tested* this?  Doesn't this basically do a dlclose() of the legacy.so subordinate libcrypto where it actually gets unloaded, and then a dlopen() of that same subordinate libcrypto?  Shouldn't all BSS and DATA areas be cleared / reset to their original values then?

mattcaswell commented:
> Have you tested this? Doesn't this basically do a dlclose() of the legacy.so subordinate libcrypto where it actually gets unloaded, and then a dlopen() of that same subordinate libcrypto? Shouldn't all BSS and DATA areas be cleared / reset to their original values then?

No - not tested. This is a hypothetical situation I'm worried about. In my understanding , if you call dlopen multiple times on the same .so it effectively ref counts things, and dlclose decrements the ref count. Only when the ref count drops to zero does the .so get removed from memory (if the OS gets around to it). In the above example the dlopen ref count does not reach zero (because of the explicit DSO_load() at the beginning) and therefore the .so never gets removed from memory, even though the active provider count has dropped to zero after we free the first libctx in my example.

rsbeckerca commented:
I am reopening #17537 because this branch is being actively modified and will continue testing at least daily.

levitte commented:
> > Have you tested this? Doesn't this basically do a dlclose() of the legacy.so subordinate libcrypto where it actually gets unloaded, and then a dlopen() of that same subordinate libcrypto? Shouldn't all BSS and DATA areas be cleared / reset to their original values then?
> 
> No - not tested. This is a hypothetical situation I'm worried about. In my understanding , if you call dlopen multiple times on the same .so it effectively ref counts things, and dlclose decrements the ref count. Only when the ref count drops to zero does the .so get removed from memory (if the OS gets around to it). In the above example the dlopen ref count does not reach zero (because of the explicit DSO_load() at the beginning) and therefore the .so never gets removed from memory, even though the active provider count has dropped to zero after we free the first libctx in my example.

I've been thinking a bit about that extra DSO_load... Is it still necessary?

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mattcaswell commented:
> I've been thinking a bit about that extra DSO_load... Is it still necessary with this PR?

It's not *necessary* at all. You are not required to call DSO_load. I added it merely to demonstrate the point: the lifetime of the DSO and the lifetime of the provider in it are not necessarily the same thing. You cannot assume that because there are no active providers left that the DSO has been unloaded.

levitte commented:
> > I've been thinking a bit about that extra DSO_load... Is it still necessary with this PR?
> 
> It's not _necessary_ at all. You are not required to call DSO_load. I added it merely to demonstrate the point: the lifetime of the DSO and the lifetime of the provider in it are not necessarily the same thing. You cannot assume that because there are no active providers left that the DSO has been unloaded.

I think that what you're saying is incorrect...  I'm convinced that it's possible to have control over who shares each copy of libcrypto, possibly with some cooperation, like this PR suggests.  In essence, that extra DSO_load doesn't really disturb much, since it's cleaned up by `OPENSSL_cleanup()` anyway...  right?

mattcaswell commented:
> I think that what you're saying is incorrect... I'm convinced that it's possible to have control over who shares each copy of libcrypto, possibly with some cooperation, like this PR suggests. In essence, that extra DSO_load doesn't really disturb much, since it's cleaned up by OPENSSL_cleanup() anyway... right?

I think you are misunderstanding what I'm trying to demonstrate. The point is after the first OSSL_LIB_CTX_free() the number of active providers drops to 0 - so OPENSSL_cleanup() gets called in the provider's instance of libcrypto. But the DSO itself is still in memory. We then create a new lib ctx and load the legacy provider again. Since the DSO is still in memory nothing is reset, and so any calls to OPENSSL_init_crypto are going to behave incorrectly (OPENSSL_cleanup has already been called, and you cannot re-initialise).

mattcaswell commented:
Just to be clear this is a -1 on this PR for the time being. It might be that this is the best we can do, but I'm not quite convinced of that yet.

mattcaswell commented:
To try and make my point a little clearer I put together the hack below. This test passes in master but fails with the code in this PR applied.

Apply the patch below and then run the test like this:

````
$ cd test
$ ../util/shlib_wrap.sh ./load_prov_test
1..1
# ERROR: (ptr) 'legacy != NULL' failed @ test/load_prov_test.c:50
# 0x0
# 4067A5CD437F0000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto/provider_core.c:922:name=legacy
# OPENSSL_TEST_RAND_ORDER=1643042390
not ok 1 - test_load_prov_twice
````

With the following patch applied.

```` diff
diff --git a/test/build.info b/test/build.info
index ec4bd8d5db..8a171cc64a 100644
--- a/test/build.info
+++ b/test/build.info
@@ -62,7 +62,7 @@ IF[{- !$disabled{tests} -}]
           context_internal_test aesgcmtest params_test evp_pkey_dparams_test \
           keymgmt_internal_test hexstr_test provider_status_test defltfips_test \
           bio_readbuffer_test user_property_test pkcs7_test upcallstest \
-          provfetchtest prov_config_test rand_test
+          provfetchtest prov_config_test rand_test load_prov_test
 
   IF[{- !$disabled{'deprecated-3.0'} -}]
     PROGRAMS{noinst}=enginetest
@@ -166,6 +166,10 @@ IF[{- !$disabled{tests} -}]
   INCLUDE[evp_extra_test]=../include ../apps/include
   DEPEND[evp_extra_test]=../libcrypto.a libtestutil.a
 
+  SOURCE[load_prov_test]=load_prov_test.c
+  INCLUDE[load_prov_test]=../include ../apps/include
+  DEPEND[load_prov_test]=../libcrypto.a libtestutil.a
+
   SOURCE[evp_extra_test2]=evp_extra_test2.c
   INCLUDE[evp_extra_test2]=../include ../apps/include
   DEPEND[evp_extra_test2]=../libcrypto libtestutil.a
diff --git a/test/load_prov_test.c b/test/load_prov_test.c
new file mode 100644
index 0000000000..0f6516a6fb
--- /dev/null
+++ b/test/load_prov_test.c
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2015-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/crypto.h>
+#include <openssl/provider.h>
+#include <dlfcn.h>
+#include "testutil.h"
+
+static int test_load_prov_twice(void)
+{
+    void *legacydso = NULL;
+    OSSL_LIB_CTX *ctx = NULL;
+    OSSL_PROVIDER *legacy = NULL;
+    int ret = 0;
+
+    /* Simulate the DSO staying around for longer than we might expect. */
+    legacydso = dlopen("../providers/legacy.so", RTLD_NOW);
+    if (!TEST_ptr(legacydso))
+        goto err;
+
+    ctx = OSSL_LIB_CTX_new();
+    if (!TEST_ptr(ctx))
+        goto err;
+    legacy = OSSL_PROVIDER_load(ctx, "legacy");
+    if (!TEST_ptr(legacy))
+        goto err;
+    OSSL_PROVIDER_unload(legacy);
+    OSSL_LIB_CTX_free(ctx);
+    legacy = NULL;
+    ctx = NULL;
+
+    /*
+     * No active references to the legacy provider any more. It should have had
+     * its teardown and cleanup functions called.
+     */
+
+
+    /* Now reload legacy */
+    ctx = OSSL_LIB_CTX_new();
+    if (!TEST_ptr(ctx))
+        goto err;
+    legacy = OSSL_PROVIDER_load(ctx, "legacy");
+    if (!TEST_ptr(legacy))
+        goto err;
+
+    ret = 1;
+ err:
+    OSSL_PROVIDER_unload(legacy);
+    OSSL_LIB_CTX_free(ctx);
+    dlclose(legacydso);
+    return ret;
+}
+
+int setup_tests(void)
+{
+    if (!test_skip_common_options()) {
+        TEST_error("Error parsing test options\n");
+        return 0;
+    }
+    ADD_TEST(test_load_prov_twice);
+
+
+    return 1;
+}
````

mattcaswell commented:
Note also that internally when we unload a provider we call dlclose() (for Unix like systems). The dlclose() documentation says this:

> The use of dlclose() reflects a statement of intent on the part of the process, but does not create any requirement upon the implementation, such as removal of the code or symbols referenced by handle.

See:
https://pubs.opengroup.org/onlinepubs/000095399/functions/dlclose.html

So even though we have attempted to remove the provider from memory, we have no guarantee that that has actually happened.

levitte commented:
> So even though we have attempted to remove the provider from memory, we have no guarantee that that has actually happened.

... but you had to make something quite deliberate to cause this.  In my view, the application have that kind of control, including to deliberately botch things up.
I don't know how we'd control that, except an init/fini or constructor/destructor approach, done in libcrypto.

That being said, this doesn't take away the need for cleanup if the provider is linked with libcrypto.a...  I believe that the easiest way is an approach similar to linking a program, i.e. having to link with an object file that we build an install alongside libcrypto.a (for programs, that is - or at least was - known as crt0.o), because such an object file inside libcrypto.a doesn't necessarily get included when linking the provider, since they aren't explicitly referred to.

mattcaswell commented:
> but you had to make something quite deliberate to cause this....the application have that kind of control

The deliberate thing that I did is also something that the OS might choose to do without any input from the application. The OS is *under no obligation* to unload a .so straight away after dlclose. I've done some testing and on Linux it does seem to be the case that it does unload straight away (at least I couldn't cause it to not do that). But I would not at all be surprised if we learned that on some platforms there is some delay between the dlclose call and the subsequent unloading from memory. If that is the case then this PR would (IMO) represent a significant break on that platform.

mattcaswell commented:
> Another thought is that we might have to stop using all those run-once flags

Most of these could be pushed into OSSL_LIB_CTX. The one controlling access to the default libctx is the hardest one to deal with...since that clearly can't be in the libctx.

levitte commented:
> > but you had to make something quite deliberate to cause this....the application have that kind of control
> 
> The deliberate thing that I did is also something that the OS might choose to do without any input from the application. The OS is _under no obligation_ to unload a .so straight away after dlclose. I've done some testing and on Linux it does seem to be the case that it does unload straight away (at least I couldn't cause it to not do that). But I would not at all be surprised if we learned that on some platforms there is some delay between the dlclose call and the subsequent unloading from memory. If that is the case then this PR would (IMO) represent a significant break on that platform.

... and that still doesn't answer my question.

levitte commented:
> ... and that still doesn't answer my question.

I think I finally have an idea how to actually test what I'm after.

levitte commented:
> > ... and that still doesn't answer my question.
> 
> I think I finally have an idea how to actually test what I'm after.

...  and I did, and sadly, I must admit that it looks like @mattcaswell was right.  I'll submit the hack I did to check this as an experimental PR

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 309 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 340 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 371 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 402 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 433 days ago

tom-cosgrove-arm commented:
Look likes there are conflicts to be addressed if this review is still wanted

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

paulidale commented:
I think this should be closed and an entirely different clean up mechanism implemented.  One that depends on the calling application shutting things down nicely.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 309 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 340 days ago

t8m commented:
@levitte Should we close this or do you plan to revive the PR at some point?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR has been closed.  It was waiting for the creator to make requested changes but it has not been updated for 90 days.
